File: templates/storage/python/server/storage.py
"""The S3 client, and the bucket it expects to find.

S3 here is the protocol: the same code reaches MinIO, Cloudflare R2 or S3
itself, and which one is decided by S3_ENDPOINT in .env.
"""

import os
import time

import boto3
from botocore.exceptions import BotoCoreError, ClientError

BUCKET = os.getenv("S3_BUCKET", "__BUCKET_NAME__")

# Only S3-compatible stores need an endpoint; S3 itself is found from the
# region. Emptying S3_ENDPOINT is what moves you from MinIO to AWS.
ENDPOINT = os.getenv("S3_ENDPOINT") or None

s3 = boto3.client(
    "s3",
    endpoint_url=ENDPOINT,
    region_name=os.getenv("AWS_REGION", "us-east-1"),
)


def ensure_bucket() -> None:
    """Creates the bucket unless it is already there.

    The retries are for the seconds after `npm run docker:up` in which MinIO is
    running but not yet answering.
    """
    for attempt in range(5):
        try:
            s3.head_bucket(Bucket=BUCKET)
            return
        except ClientError as error:
            if error.response["Error"]["Code"] in ("404", "NoSuchBucket"):
                s3.create_bucket(Bucket=BUCKET)
                return
            if attempt == 4:
                raise
        except BotoCoreError:
            if attempt == 4:
                raise
        time.sleep(1)

File: templates/storage/typescript/server/src/storage.ts
import { CreateBucketCommand, HeadBucketCommand, S3Client } from "@aws-sdk/client-s3";

export const BUCKET = process.env.S3_BUCKET ?? "__BUCKET_NAME__";

// Only S3-compatible stores need an endpoint; S3 itself is found from the
// region. Emptying S3_ENDPOINT in .env is what moves you from MinIO to AWS.
const endpoint = process.env.S3_ENDPOINT || undefined;

export const s3 = new S3Client({
  region: process.env.AWS_REGION ?? "us-east-1",
  endpoint,
  // MinIO addresses a bucket as a path, where AWS uses a subdomain.
  forcePathStyle: Boolean(endpoint),
});

/**
 * Creates the bucket unless it is already there, the way the database layers
 * create their tables on startup. The retries are for the seconds after
 * `npm run docker:up` in which MinIO is running but not yet answering.
 */
export async function ensureBucket(): Promise<void> {
  for (let attempt = 1; attempt <= 5; attempt++) {
    try {
      await s3.send(new HeadBucketCommand({ Bucket: BUCKET }));
      return;
    } catch (error) {
      const status = (error as { $metadata?: { httpStatusCode?: number } }).$metadata
        ?.httpStatusCode;
      if (status === 404) {
        await s3.send(new CreateBucketCommand({ Bucket: BUCKET }));
        return;
      }
      if (attempt === 5) throw error;
      await new Promise((resolve) => setTimeout(resolve, 1000));
    }
  }
}
